File: VortexVise/Player.cpp
#include "Player.h"

void Player::ProcessInput(float deltaTime, Input input)
{
	// TODO: NETWORKING
	// Physics Runs On The Server
	/*
	struct Input
    {
         bool left;
         bool right;
         bool forward;
         bool back;
         bool jump;
    };

    struct State
    {
         Vector position;
         Vector velocity;
    };
	 struct Move
    {
        double time;
        Input input;
        State state;
    };
	*/

	// I should receive the input here, this should happen on the server and the client should have predictions, the later the server responds
	if (input.Right) {
		m_velocity.x += m_acceleration * deltaTime;
		if (m_velocity.x > m_maxMoveSpeed)// && gravitationalForce == 0)
			m_velocity.x = m_maxMoveSpeed;
		m_direction = -1;
	}
	else if (input.Left) {
		m_velocity.x -= m_acceleration * deltaTime;
		if (m_velocity.x < m_maxMoveSpeed * -1)// && gravitationalForce == 0)
			m_velocity.x = m_maxMoveSpeed * -1;
		m_direction = 1;
	}
	else {
		float desaceleration = m_isTouchingTheGround || m_velocity.y == 0 ? 10 : 0.5;
		m_velocity.x = Lerp(m_velocity.x, 0, 1 - expf(-desaceleration * deltaTime));
	}

	if (m_velocity.x != 0)
		m_position.x += m_velocity.x * deltaTime;

	//if (IsKeyDown(KEY_SPACE)
	if (input.Jump && m_isTouchingTheGround) {
		m_isTouchingTheGround = false;
		m_velocity.y = -400;
	}

}

Input Player::GetInput()
{
	// TODO: Implement gamepad and stuff
	Input input{};
	if (IsKeyDown(KEY_A))
		input.Left = true;
	if (IsKeyDown(KEY_D))
		input.Right = true;
	if (IsKeyDown(KEY_SPACE))
		input.Jump = true;
	if (IsKeyPressed(KEY_SPACE))
		input.CancelHook = true;
	if (IsMouseButtonDown(1))
		input.Hook = true;
	if (IsKeyDown(KEY_W))
		input.Up = true;
	if (IsKeyDown(KEY_S))
		input.Down = true;
	return input;
}

void Player::ApplyGravitationalForce(float gravity, float deltaTime)
{
	float maxGravity = 500;
	if (!m_isTouchingTheGround) {
		m_velocity.y += gravity * deltaTime;
		if (m_velocity.y >= maxGravity)
			m_velocity.y = maxGravity;
		m_position.y += m_velocity.y * deltaTime;
	}
}

Vector2 Player::GetPosition() const
{
	return m_position;
}

float Player::GetX() const
{
	return m_position.x;
}

float Player::GetY() const
{
	return m_position.y;
}

float Player::GetGravitationalForce() const
{
	return m_velocity.y;
}

float Player::GetMoveSpeed() const
{
	return m_velocity.x;
}

void Player::ApplyCollisions(Map& map)
{
	Vector2 collisionOffset = { 20,12 };
	Rectangle endingCollision = { m_position.x + collisionOffset.x,m_position.y + collisionOffset.y,25,40 };
	m_isTouchingTheGround = false;

	Vector2 mapSize = map.GetMapSize();
	// Apply ouside map collisions
	if (endingCollision.y <= 0) {
		m_position.y = 11.9;
		m_velocity.y = 0;
	}
	else if (endingCollision.y > mapSize.y) {
		// TODO: Kill the player
		m_position = { map.GetMapSize().x / 2,map.GetMapSize().y / 2 };
		m_velocity.y = 0;
		m_velocity.x = 0;

	}
	if (endingCollision.x <= 0) {
		m_position.x = 0 - (endingCollision.x - m_position.x);
		m_velocity.x = 0;
	}
	else if (endingCollision.x + endingCollision.width >= mapSize.x) {
		m_position.x = mapSize.x - endingCollision.width - collisionOffset.x;
		m_velocity.x = 0;
	}

	// This will interpolate the collisions when the player is fast, otherwise he will go through stuff easily
	// WARNING: This solution only works if the player never goes in the minus coordinates, why? because at least for now he can't, if this changes please redo this collision interpolation crap
	m_playerCollisions.clear();
	float interpolationAmount = 2;
	for (float i = interpolationAmount; i > 0; i -= 0.2) {
		Rectangle interpolatedCollision = endingCollision;
		if (m_collisionBox.x < endingCollision.x && endingCollision.x - m_collisionBox.x >= m_collisionBox.width * i) {
			interpolatedCollision.x += m_collisionBox.width * i;
		}
		else if (m_collisionBox.x > endingCollision.x && m_collisionBox.x - endingCollision.x >= m_collisionBox.width * i) {
			interpolatedCollision.x -= m_collisionBox.width * i;
		}

		if (m_collisionBox.y < endingCollision.y && endingCollision.y - m_collisionBox.y >= m_collisionBox.height * i) {
			interpolatedCollision.y += m_collisionBox.height * i;
		}
		else if (m_collisionBox.y > endingCollision.y && m_collisionBox.y - endingCollision.y >= m_collisionBox.height * i) {
			interpolatedCollision.y -= m_collisionBox.height * i;
		}
		if (interpolatedCollision.x != endingCollision.x || interpolatedCollision.y != endingCollision.y)
			m_playerCollisions.push_front(interpolatedCollision);
	}


	m_playerCollisions.push_back(endingCollision);

	// Apply map collisions
	for (auto& playerCollision : m_playerCollisions) {
		bool colided = false;
		for (const auto& collision : map.GetCollisions())
		{
			if (CheckCollisionRecs(playerCollision, collision)) {

				// This means the player is inside the thing 
				auto collisionOverlap = GetCollisionRec(playerCollision, collision);

				if (m_position.y == collision.y - m_texture.height + collisionOffset.y)
					m_isTouchingTheGround = true;

				Vector2 colliderCenter = { collision.x + collision.width * 0.5, collision.y + collision.height * 0.5 };

				if (collisionOverlap.height < collisionOverlap.width) {
					if (collisionOverlap.y < colliderCenter.y) {
						// Feet collision
						m_position.y = collision.y - m_texture.height + collisionOffset.y;
						playerCollision.y = (collision.y - playerCollision.height);
						m_velocity.y = 0;
						m_isTouchingTheGround = true;
						m_collisionBox = playerCollision;
						colided = true;
						continue;
					}
					else {
						// Head collision
						//m_position.y = collision.y + collision.height + collisionOffset.y;
						//playerCollision.y = collision.y + collision.height;
						m_position.y += collisionOverlap.height;
						playerCollision.y += collisionOverlap.height;
						m_velocity.y = 0.01;
						m_collisionBox = playerCollision;
						colided = true;
						continue;
					}
				}
				else {

					if (collisionOverlap.x > colliderCenter.x) {
						m_velocity.x = 0;
						// Right side of collision block on map
						m_position.x += collisionOverlap.width;
						playerCollision.x += collisionOverlap.width;
						m_collisionBox = playerCollision;
						colided = true;
						continue;
					}
					else
					{
						m_velocity.x = 0;
						// Left collision
						m_position.x -= collisionOverlap.width;
						playerCollision.x -= collisionOverlap.width;
						m_collisionBox = playerCollision;
						colided = true;
						continue;
					}
				}


			}
		}

		if (colided)
			return;
	}
	m_collisionBox = endingCollision;
}

void Player::ProcessCamera(Map& map)
{
	if (m_hasCamera) {
		Vector2 target = { m_position.x ,m_position.y };

		// Make it stay inside the map
		if (target.x - GetScreenWidth() / 2 <= 0)
			target.x = GetScreenWidth() / 2;
		else if (target.x + GetScreenWidth() / 2 >= map.GetMapSize().x)
			target.x = map.GetMapSize().x - GetScreenWidth() / 2;

		if (target.y - GetScreenHeight() / 2 <= 0)
			target.y = GetScreenHeight() / 2;
		else if (target.y + GetScreenHeight() / 2 >= map.GetMapSize().y)
			target.y = map.GetMapSize().y - GetScreenHeight() / 2;

		// Make camera smooth
		// FIXME: fix camera jerkness when almost hitting the target
		m_camera.target.x = Lerp(m_camera.target.x, target.x, 1 - expf(-3 * GetFrameTime()));
		m_camera.target.y = Lerp(m_camera.target.y, target.y, 1 - expf(-3 * GetFrameTime()));

		BeginMode2D(m_camera);
	}
}

Vector2 Player::GetPlayerCenterPosition() const
{
	Vector2 position = { m_position.x, m_position.y };
	position.x += m_texture.width / 2;
	position.y += m_texture.height / 2;
	return position;
}

bool Player::IsLookingRight() const
{
	return m_direction == -1;
}

void Player::AddVelocity(Vector2 velocity, float deltaTime)
{
	m_velocity.x += velocity.x * deltaTime;
	m_velocity.y += velocity.y * deltaTime;
}

void Player::ApplyVelocity(float deltaTime)
{
	m_position.x += m_velocity.x * deltaTime;
	m_position.y += m_velocity.y * deltaTime;
}

void Player::Draw()
{
	Rectangle sourceRec = Rectangle{ 0.0f, 0.0f, (float)m_texture.width * m_direction, (float)m_texture.height };

	Rectangle destRec = Rectangle{ 0, 0, (float)m_texture.width, (float)m_texture.height };

	DrawTexturePro(m_texture, sourceRec, destRec, { m_position.x * -1,m_position.y * -1 }, 0, WHITE);


	if (Utils::Debug()) {
		DrawRectangleRec(m_collisionBox, GREEN); // Debug
		unsigned char color = 0;
		for (auto& playerCollision : m_playerCollisions) {
			DrawRectangleRec(playerCollision, { color,100,100,255 }); // Collision Interpolation
			color += 40;

		}
	}
}

File: VortexVise/GameObjects/Player.cs
using Raylib_cs;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;
using VortexVise.Models;
using VortexVise.Networking;
using VortexVise.Utilities;

namespace VortexVise.GameObjects;

public class Player
{
    public Guid Id { get; private set; } = Guid.NewGuid();
    public Hook Hook { get; private set; } = new Hook();

    private Vector2 _position = new Vector2() { X = 0, Y = 0 };
    private Vector2 _velocity = new Vector2() { X = 0, Y = 0 };
    private int _direction = 1;
    private Texture2D _texture;
    private float _maxMoveSpeed = 350;
    private float _acceleration = 750;
    private Camera2D _camera;
    private bool _hasCamera = false;
    private bool _isTouchingTheGround = false;
    private Rectangle _collisionBox;
    private List<Rectangle> _playerCollisions = new List<Rectangle>();

    public Player(bool hasCamera, Map map)
    {
        _texture = Raylib.LoadTexture("Resources/Sprites/Skins/fatso.png"); // TODO: make load skin, not this hardcoded crap
        var spawnPoint = new Vector2(Raylib.GetScreenWidth() / 2.0f, Raylib.GetScreenHeight() / 2.0f);
        _position = spawnPoint;

        if (hasCamera)
        {
            _hasCamera = hasCamera;
            _camera = new Camera2D(spawnPoint, spawnPoint, 0, 1);
        }
    }

    public Vector2 ProcessVelocity(float deltaTime, Input input)
    {
        var velocity = _velocity;
        if (input.Right)
        {
            velocity.X += _acceleration * deltaTime;
            if (velocity.X > _maxMoveSpeed)// && gravitationalForce == 0)
                velocity.X = _maxMoveSpeed;
            _direction = -1;
        }
        else if (input.Left)
        {
            velocity.X -= _acceleration * deltaTime;
            if (velocity.X < _maxMoveSpeed * -1)// && gravitationalForce == 0)
                velocity.X = _maxMoveSpeed * -1;
            _direction = 1;
        }
        else
        {
            float desaceleration = _isTouchingTheGround || velocity.Y == 0f ? 10f : 0.5f;
            velocity.X = Raymath.Lerp(velocity.X, 0, 1f - (float)Math.Exp(-desaceleration * deltaTime));
        }


        if (input.Jump && _isTouchingTheGround)
        {
            _isTouchingTheGround = false;
            velocity.Y = -400;
        }

        return velocity;
    }

    public Input GetInput()
    {
        // TODO: Implement gamepad and stuff
        Input input = new();
        if (Raylib.IsKeyDown(KeyboardKey.A))
            input.Left = true;
        if (Raylib.IsKeyDown(KeyboardKey.D))
            input.Right = true;
        if (Raylib.IsKeyDown(KeyboardKey.Space) || Raylib.IsKeyDown(KeyboardKey.K))
            input.Jump = true;
        if (Raylib.IsKeyPressed(KeyboardKey.Space))
            input.CancelHook = true;
        if (Raylib.IsMouseButtonDown(MouseButton.Right) || Raylib.IsKeyDown(KeyboardKey.J))
            input.Hook = true;
        if (Raylib.IsKeyDown(KeyboardKey.W))
            input.Up = true;
        if (Raylib.IsKeyDown(KeyboardKey.S))
            input.Down = true;
        return input;
    }

    public Vector2 ProcessPosition(float gravity, float deltaTime, Vector2 velocity)
    {
        var position = _position;
        if (velocity.X != 0)
            position.X += velocity.X * deltaTime;

        float maxGravity = 500;
        if (!_isTouchingTheGround)
        {
            velocity.Y += gravity * deltaTime;
            if (velocity.Y >= maxGravity)
                velocity.Y = maxGravity;
        }
        position.Y += velocity.Y * deltaTime;
        return position;
    }

    public Vector2 GetPosition()
    {
        return _position;
    }

    public Vector2 GetVelocity()
    {
        return _velocity;
    }

    public float GetX()
    {
        return _position.X;
    }

    public float GetY()
    {
        return _position.Y;
    }

    public float GetGravitationalForce()
    {
        return _velocity.Y;
    }

    public float GetMoveSpeed()
    {
        return _velocity.X;
    }

    public void ApplyCollisions(Map map)
    {
        Vector2 collisionOffset = new(20, 12);
        Rectangle endingCollision = new(_position.X + collisionOffset.X, _position.Y + collisionOffset.Y, 25, 40);
        _isTouchingTheGround = false;

        Vector2 mapSize = map.GetMapSize();
        // Apply ouside map collisions
        if (endingCollision.Y <= 0)
        {
            _position.Y = 11.9f;
            _velocity.Y = 0;
        }
        else if (endingCollision.Y > mapSize.Y)
        {
            // TODO: Kill the player
            _position = new(map.GetMapSize().X / 2, map.GetMapSize().Y / 2);
            _velocity.Y = 0;
            _velocity.X = 0;

        }
        if (endingCollision.X <= 0)
        {
            _position.X = 0 - (endingCollision.X - _position.X);
            _velocity.X = 0;
        }
        else if (endingCollision.X + endingCollision.Width >= mapSize.X)
        {
            _position.X = mapSize.X - endingCollision.Width - collisionOffset.X;
            _velocity.X = 0;
        }

        // This will interpolate the collisions when the player is fast, otherwise he will go through stuff easily
        // WARNING: This solution only works if the player never goes in the minus coordinates, why? because at least for now he can't, if this changes please redo this collision interpolation crap
        _playerCollisions.Clear();
        float interpolationAmount = 2;
        for (float i = interpolationAmount; i > 0; i -= 0.2f)
        {
            Rectangle interpolatedCollision = endingCollision;
            if (_collisionBox.X < endingCollision.X && endingCollision.X - _collisionBox.X >= _collisionBox.Width * i)
            {
                interpolatedCollision.X += _collisionBox.Width * i;
            }
            else if (_collisionBox.X > endingCollision.X && _collisionBox.X - endingCollision.X >= _collisionBox.Width * i)
            {
                interpolatedCollision.X -= _collisionBox.Width * i;
            }

            if (_collisionBox.Y < endingCollision.Y && endingCollision.Y - _collisionBox.Y >= _collisionBox.Height * i)
            {
                interpolatedCollision.Y += _collisionBox.Height * i;
            }
            else if (_collisionBox.Y > endingCollision.Y && _collisionBox.Y - endingCollision.Y >= _collisionBox.Height * i)
            {
                interpolatedCollision.Y -= _collisionBox.Height * i;
            }
            if (interpolatedCollision.X != endingCollision.X || interpolatedCollision.Y != endingCollision.Y)
                _playerCollisions.Add(interpolatedCollision);
        }


        _playerCollisions.Add(endingCollision);

        // Apply map collisions
        foreach (var playerCollision in _playerCollisions)
        {
            bool colided = false;
            foreach (var collision in map.GetCollisions())
            {
                if (Raylib.CheckCollisionRecs(playerCollision, collision))
                {

                    // This means the player is inside the thing 
                    var collisionOverlap = Raylib.GetCollisionRec(playerCollision, collision);

                    if (_position.Y == collision.Y - _texture.Height + collisionOffset.Y)
                        _isTouchingTheGround = true;

                    Vector2 colliderCenter = new(collision.X + collision.Width * 0.5f, collision.Y + collision.Height * 0.5f);

                    if (collisionOverlap.Height < collisionOverlap.Width)
                    {
                        if (collisionOverlap.Y < colliderCenter.Y)
                        {
                            // Feet collision
                            _position.Y = collision.Y - _texture.Height + collisionOffset.Y;
                            _collisionBox = playerCollision;
                            _collisionBox.Y = (collision.Y - playerCollision.Height);
                            _velocity.Y = 0;
                            _isTouchingTheGround = true;
                            colided = true;
                            continue;
                        }
                        else
                        {
                            // Head collision
                            //_position.Y = collision.Y + collision.height + collisionOffset.Y;
                            //playerCollision.Y = collision.Y + collision.height;
                            _position.Y += collisionOverlap.Height;
                            _collisionBox = playerCollision;
                            _collisionBox.Y += collisionOverlap.Height;
                            _velocity.Y = 0.01f;
                            colided = true;
                            continue;
                        }
                    }
                    else
                    {

                        if (collisionOverlap.X > colliderCenter.X)
                        {
                            _velocity.X = 0;
                            // Right side of collision block on map
                            _position.X += collisionOverlap.Width;
                            _collisionBox = playerCollision;
                            _collisionBox.X += collisionOverlap.Width;
                            colided = true;
                            continue;
                        }
                        else
                        {
                            _velocity.X = 0;
                            // Left collision
                            _position.X -= collisionOverlap.Width;
                            _collisionBox = playerCollision;
                            _collisionBox.X -= collisionOverlap.Width;
                            colided = true;
                            continue;
                        }
                    }


                }
            }

            if (colided)
                return;
        }
        _collisionBox = endingCollision;
    }

    public void ProcessCamera(Map map)
    {
        if (_hasCamera)
        {
            Vector2 target = new(_position.X, _position.Y);

            // Make it stay inside the map
            if (target.X - Raylib.GetScreenWidth() / 2 <= 0)
                target.X = Raylib.GetScreenWidth() / 2;
            else if (target.X + Raylib.GetScreenWidth() / 2 >= map.GetMapSize().X)
                target.X = map.GetMapSize().X - Raylib.GetScreenWidth() / 2;

            if (target.Y - Raylib.GetScreenHeight() / 2 <= 0)
                target.Y = Raylib.GetScreenHeight() / 2;
            else if (target.Y + Raylib.GetScreenHeight() / 2 >= map.GetMapSize().Y)
                target.Y = map.GetMapSize().Y - Raylib.GetScreenHeight() / 2;

            // Make camera smooth
            // FIXME: fix camera jerkness when almost hitting the target
            _camera.Target.X = Raymath.Lerp(_camera.Target.X, target.X, 1 - (float)Math.Exp(-3 * Raylib.GetFrameTime()));
            _camera.Target.Y = Raymath.Lerp(_camera.Target.Y, target.Y, 1 - (float)Math.Exp(-3 * Raylib.GetFrameTime()));

            Raylib.BeginMode2D(_camera);
        }
    }

    public Vector2 GetPlayerCenterPosition()
    {

        Vector2 position = new(_position.X, _position.Y);
        position.X += _texture.Width / 2;
        position.Y += _texture.Height / 2;
        return position;
    }

    public bool IsLookingRight()
    {
        return _direction == -1;
    }

    public void AddVelocity(Vector2 velocity, float deltaTime)
    {
        _velocity.X += velocity.X * deltaTime;
        _velocity.Y += velocity.Y * deltaTime;
    }

    public void Draw()
    {
        Rectangle sourceRec = new(0.0f, 0.0f, (float)_texture.Width * _direction, (float)_texture.Height);

        Rectangle destRec = new(0, 0, (float)_texture.Width, (float)_texture.Height);

        Raylib.DrawTexturePro(_texture, sourceRec, destRec, new Vector2(_position.X * -1, _position.Y * -1), 0, Color.White);


        if (Utils.Debug())
        {
            Raylib.DrawRectangleRec(_collisionBox, Color.Green); // Debug
            int color = 0;
            foreach (var playerCollision in _playerCollisions)
            {
                Raylib.DrawRectangleRec(playerCollision, new(color, 100, 100, 255)); // Collision Interpolation
                color += 40;

            }
        }
    }
    public void ApplyState(PlayerState state)
    {
        _position = state.Position;
        _velocity = state.Velocity;
        Hook.ApplyState(state.HookState);
    }


}
